SSR with blueprint

Server side rendering with blueprint breaks, `dom4.max.js` is requiring tons of things.

Sorry for that megapost. A few more comments: 

@justingreenberg the popover/portal code is fine. As mentioned by @abhishiv, this is called from `componentDidMount`, which is never called during server rendering. 
@justingreenberg we should see to it that calls such as the one to `document.createElement` do not run on the server. In doing so we can avoid having to shim `document` on the server. The calls to `createElement` also don't create elements in the React-managed part of the DOM. With this in mind, there should be nothing to cause a checksum mismatch and re-render on the client. 

@RobAWilkinson @schrepfler [DOM4](https://github.com/WebReflection/dom4) is a polyfill for the [W3C DOM level 4 living specification](https://www.w3.org/TR/dom/). Because this is related to DOM manipulation, it isn't needed server-side (but is still be needed browser-side).

@isTravis that's an interesting approach, but seems overcomplicated given that the amount of patching needed to get Blueprintjs to work server-side is minimal.     

@justingreenberg your patch for conditional import of 'dom4' breaks the build 
```
packages\core\src\components\overlay\overlay.tsx(301,55): error TS2339: Property 'query' does not exist on type 'HTMLElement'.
```
unless the dom4 types (required for TypeScript to correctly compile) are included. This is an easy change, though:
```
===================================================================
--- packages/core/package.json	(revision 497a03eab4420e48db0865407995895aa25726fa)
+++ packages/core/package.json	(revision )
@@ -6,6 +6,7 @@
   "typings": "dist/index.d.ts",
   "style": "dist/blueprint.css",
   "dependencies": {
+    "@types/dom4": "^1.5.20",
     "classnames": "^2.2",
     "dom4": "^1.8",
     "normalize.css": "4.1.1",
```

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -6,6 +6,7 @@
   "typings": "dist/index.d.ts",
   "style": "dist/blueprint.css",
   "dependencies": {
+    "@types/dom4": "^1.5.20",
     "classnames": "^2.2",
     "dom4": "^1.8",
     "normalize.css": "4.1.1",

--- a/packages/core/src/accessibility/focusStyleManager.ts
+++ b/packages/core/src/accessibility/focusStyleManager.ts
@@ -9,7 +9,7 @@ import { InteractionModeEngine } from "../common/interactionMode";
 
 export const FOCUS_DISABLED_CLASS = "pt-focus-disabled";
 
-const focusEngine = new InteractionModeEngine(document.documentElement, FOCUS_DISABLED_CLASS);
+const focusEngine = new InteractionModeEngine(typeof document != 'undefined' ? document.documentElement : null, FOCUS_DISABLED_CLASS);
 
 // this is basically meaningless to unit test; it requires manual UI testing
 /* istanbul ignore next */

--- a/packages/core/src/accessibility/focusStyleManager.ts
+++ b/packages/core/src/accessibility/focusStyleManager.ts
@@ -9,7 +9,14 @@ import { InteractionModeEngine } from "../common/interactionMode";
 
 export const FOCUS_DISABLED_CLASS = "pt-focus-disabled";
 
-const focusEngine = new InteractionModeEngine(typeof document != 'undefined' ? document.documentElement : null, FOCUS_DISABLED_CLASS);
+const fakeFocusEngine = {
+    stop: () => true,
+    isActive: () => true,
+    start: () => true
+};
+const focusEngine = typeof document != 'undefined'
+    ? new InteractionModeEngine(document.documentElement, FOCUS_DISABLED_CLASS)
+    : fakeFocusEngine;
 
 // this is basically meaningless to unit test; it requires manual UI testing
 /* istanbul ignore next */

--- a/packages/core/src/accessibility/focusStyleManager.ts
+++ b/packages/core/src/accessibility/focusStyleManager.ts
@@ -4,6 +4,7 @@
  * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
  * and https://github.com/palantir/blueprint/blob/master/PATENTS
  */
+/* istanbul ignore next */
 
 import { InteractionModeEngine } from "../common/interactionMode";
 

--- a/packages/core/src/accessibility/focusStyleManager.ts
+++ b/packages/core/src/accessibility/focusStyleManager.ts
@@ -4,17 +4,19 @@
  * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
  * and https://github.com/palantir/blueprint/blob/master/PATENTS
  */
-/* istanbul ignore next */
 
 import { InteractionModeEngine } from "../common/interactionMode";
 
 export const FOCUS_DISABLED_CLASS = "pt-focus-disabled";
 
+/* istanbul ignore next */
 const fakeFocusEngine = {
     isActive: () => true,
     start: () => true,
     stop: () => true,
 };
+
+/* istanbul ignore next */
 const focusEngine = typeof document !== "undefined"
     ? new InteractionModeEngine(document.documentElement, FOCUS_DISABLED_CLASS)
     : fakeFocusEngine;

--- a/packages/core/src/accessibility/focusStyleManager.ts
+++ b/packages/core/src/accessibility/focusStyleManager.ts
@@ -10,11 +10,11 @@ import { InteractionModeEngine } from "../common/interactionMode";
 export const FOCUS_DISABLED_CLASS = "pt-focus-disabled";
 
 const fakeFocusEngine = {
-    stop: () => true,
     isActive: () => true,
-    start: () => true
+    start: () => true,
+    stop: () => true,
 };
-const focusEngine = typeof document != 'undefined'
+const focusEngine = typeof document !== "undefined"
     ? new InteractionModeEngine(document.documentElement, FOCUS_DISABLED_CLASS)
     : fakeFocusEngine;
 

--- a/packages/core/src/components/hotkeys/hotkeyParser.ts
+++ b/packages/core/src/components/hotkeys/hotkeyParser.ts
@@ -112,7 +112,7 @@ export const Aliases = {
     command: "meta",
     escape: "esc",
     minus: "-",
-    mod: ((typeof navigator != 'undefined') && /Mac|iPod|iPhone|iPad/.test(navigator.platform)) ? "meta" : "ctrl",
+    mod: ((typeof navigator !== "undefined") && /Mac|iPod|iPhone|iPad/.test(navigator.platform)) ? "meta" : "ctrl",
     option: "alt",
     plus: "+",
     return: "enter",

--- a/packages/core/src/components/hotkeys/hotkeyParser.ts
+++ b/packages/core/src/components/hotkeys/hotkeyParser.ts
@@ -112,7 +112,7 @@ export const Aliases = {
     command: "meta",
     escape: "esc",
     minus: "-",
-    mod: /Mac|iPod|iPhone|iPad/.test(navigator.platform) ? "meta" : "ctrl",
+    mod: ((typeof navigator != 'undefined') && /Mac|iPod|iPhone|iPad/.test(navigator.platform)) ? "meta" : "ctrl",
     option: "alt",
     plus: "+",
     return: "enter",

--- a/packages/core/src/components/hotkeys/hotkeysEvents.ts
+++ b/packages/core/src/components/hotkeys/hotkeysEvents.ts
@@ -13,7 +13,7 @@ import { comboMatches, getKeyCombo, IKeyCombo, parseKeyCombo } from "./hotkeyPar
 import { IHotkeysProps } from "./hotkeys";
 import { isHotkeysDialogShowing, showHotkeysDialog } from "./hotkeysDialog";
 
-const SHOW_DIALOG_KEY_COMBO = parseKeyCombo("?");
+const SHOW_DIALOG_KEY = "?";
 
 export enum HotkeyScope {
     LOCAL,
@@ -59,7 +59,7 @@ export class HotkeysEvents {
 
         const combo = getKeyCombo(e);
 
-        if (comboMatches(SHOW_DIALOG_KEY_COMBO, combo)) {
+        if (comboMatches(parseKeyCombo(SHOW_DIALOG_KEY), combo)) {
             showHotkeysDialog(this.actions.map((action) => action.props));
             return;
         }

--- a/packages/core/src/components/hotkeys/hotkeysTarget.tsx
+++ b/packages/core/src/components/hotkeys/hotkeysTarget.tsx
@@ -48,7 +48,6 @@ export function HotkeysTarget<T extends { prototype: IHotkeysTarget }>(construct
         }
     };
 
-    // tslint:disable no-invalid-this only-arrow-functions
     constructor.prototype.componentDidMount = function() {
         // attach global key event listeners
         document.addEventListener("keydown", this.globalHotkeysEvents.handleKeyDown);

--- a/packages/core/src/components/hotkeys/hotkeysTarget.tsx
+++ b/packages/core/src/components/hotkeys/hotkeysTarget.tsx
@@ -26,7 +26,13 @@ export interface IHotkeysTarget extends React.Component<any, any>, React.Compone
 }
 
 export function HotkeysTarget<T extends { prototype: IHotkeysTarget }>(constructor: T) {
-    const { componentWillMount, componentDidMount, componentWillUnmount, render, renderHotkeys } = constructor.prototype;
+    const {
+        componentWillMount,
+        componentDidMount,
+        componentWillUnmount,
+        render,
+        renderHotkeys,
+    } = constructor.prototype;
 
     if (!isFunction(renderHotkeys)) {
         throw new Error(`@HotkeysTarget-decorated class must implement \`renderHotkeys\`. ${constructor}`);

--- a/packages/core/src/components/hotkeys/hotkeysTarget.tsx
+++ b/packages/core/src/components/hotkeys/hotkeysTarget.tsx
@@ -26,7 +26,7 @@ export interface IHotkeysTarget extends React.Component<any, any>, React.Compone
 }
 
 export function HotkeysTarget<T extends { prototype: IHotkeysTarget }>(constructor: T) {
-    const { componentWillMount, componentWillUnmount, render, renderHotkeys } = constructor.prototype;
+    const { componentWillMount, componentDidMount, componentWillUnmount, render, renderHotkeys } = constructor.prototype;
 
     if (!isFunction(renderHotkeys)) {
         throw new Error(`@HotkeysTarget-decorated class must implement \`renderHotkeys\`. ${constructor}`);
@@ -37,12 +37,19 @@ export function HotkeysTarget<T extends { prototype: IHotkeysTarget }>(construct
         this.localHotkeysEvents = new HotkeysEvents(HotkeyScope.LOCAL);
         this.globalHotkeysEvents = new HotkeysEvents(HotkeyScope.GLOBAL);
 
+        if (componentWillMount != null) {
+            componentWillMount.call(this);
+        }
+    };
+
+    // tslint:disable no-invalid-this only-arrow-functions
+    constructor.prototype.componentDidMount = function() {
         // attach global key event listeners
         document.addEventListener("keydown", this.globalHotkeysEvents.handleKeyDown);
         document.addEventListener("keyup", this.globalHotkeysEvents.handleKeyUp);
 
-        if (componentWillMount != null) {
-            componentWillMount.call(this);
+        if (componentDidMount != null) {
+            componentDidMount.call(this);
         }
     };
 

--- a/packages/core/src/components/index.ts
+++ b/packages/core/src/components/index.ts
@@ -5,7 +5,11 @@
  * and https://github.com/palantir/blueprint/blob/master/PATENTS
  */
 
-import "dom4";
+declare function require(moduleName: string): any; //declare node.js 'require' so that we can conditionally import
+if (typeof window != 'undefined' && typeof document != 'undefined') { //we're in browser
+    require("dom4"); //only import actual dom4 if we're in the browser (not server-compatible)
+    //we'll still need dom4 types for the TypeScript to compile, these are included in package.json
+}
 
 import * as contextMenu from "./context-menu/contextMenu";
 export const ContextMenu = contextMenu;

--- a/packages/core/src/components/index.ts
+++ b/packages/core/src/components/index.ts
@@ -5,10 +5,11 @@
  * and https://github.com/palantir/blueprint/blob/master/PATENTS
  */
 
-declare function require(moduleName: string): any; //declare node.js 'require' so that we can conditionally import
-if (typeof window != 'undefined' && typeof document != 'undefined') { //we're in browser
-    require("dom4"); //only import actual dom4 if we're in the browser (not server-compatible)
-    //we'll still need dom4 types for the TypeScript to compile, these are included in package.json
+declare function require(moduleName: string): any; // declare node.js "require" so that we can conditionally import
+if (typeof window !== "undefined" && typeof document !== "undefined") { // we're in browser
+    // tslint:disable-next-line:no-var-requires
+    require("dom4"); // only import actual dom4 if we're in the browser (not server-compatible)
+    // we'll still need dom4 types for the TypeScript to compile, these are included in package.json
 }
 
 import * as contextMenu from "./context-menu/contextMenu";

--- a/packages/core/src/components/menu/menuItem.tsx
+++ b/packages/core/src/components/menu/menuItem.tsx
@@ -176,8 +176,8 @@ export class MenuItem extends AbstractComponent<IMenuItemProps, IMenuItemState> 
             }
 
             let { left = 0, right = 0 } = this.props.submenuViewportMargin;
-            if (typeof document != 'undefined'
-                && typeof document.documentElement != 'undefined'
+            if (typeof document !== "undefined"
+                && typeof document.documentElement !== "undefined"
                 && Number(document.documentElement.clientWidth)) {
                 // we're in a browser context and the clientWidth is available,
                 // use it to set calculate 'right'

--- a/packages/core/src/components/menu/menuItem.tsx
+++ b/packages/core/src/components/menu/menuItem.tsx
@@ -176,7 +176,13 @@ export class MenuItem extends AbstractComponent<IMenuItemProps, IMenuItemState> 
             }
 
             let { left = 0, right = 0 } = this.props.submenuViewportMargin;
-            right = document.documentElement.clientWidth - right;
+            if (typeof document != 'undefined'
+                && typeof document.documentElement != 'undefined'
+                && Number(document.documentElement.clientWidth)) {
+                // we're in a browser context and the clientWidth is available,
+                // use it to set calculate 'right'
+                right = document.documentElement.clientWidth - right;
+            }
             // uses context to prioritize the previous positioning
             let alignLeft = this.context.alignLeft || false;
             if (alignLeft) {

